src/develop/pixelpipe_hb.c

```diff
@@ -2547,6 +2564,142 @@ float *dt_dev_get_raster_mask(const dt_dev_pixelpipe_t *pipe, const dt_iop_modul
   return raster_mask;
 }
 
+void dt_dev_clear_luminance_mask(dt_dev_pixelpipe_t *pipe)
+{
+  if(pipe->luminance_mask_data) dt_free_align(pipe->luminance_mask_data);
+  pipe->luminance_mask_data = NULL;
+}
+
+gboolean dt_dev_write_luminance_mask(dt_dev_pixelpipe_iop_t *piece, float *const rgb, const dt_iop_roi_t *const roi_in, const int mode)
+{
+  dt_dev_pixelpipe_t *p = piece->pipe;
+  if((p->want_luminance_mask & DT_DEV_LUMINANCE_MASK_REQUIRED) == 0) return FALSE;
+  if((p->want_luminance_mask & ~DT_DEV_LUMINANCE_MASK_REQUIRED) != mode) return FALSE;
+
+  dt_dev_clear_luminance_mask(p);
+
+  const int width = roi_in->width;
+  const int height = roi_in->height;
+  float *mask = dt_alloc_align_float((size_t)width * height);
+  if(mask == NULL) return TRUE;
+
+  p->luminance_mask_data = mask;
+  memcpy(&p->luminance_mask_roi, roi_in, sizeof(dt_iop_roi_t));
+
+  dt_masks_calc_luminance_mask(rgb, mask, width, height);
+  return FALSE;
+}
+
+#ifdef HAVE_OPENCL
+gboolean dt_dev_write_luminance_mask_cl(dt_dev_pixelpipe_iop_t *piece, cl_mem in, const dt_iop_roi_t *const roi_in, const int mode)
+{
+  dt_dev_pixelpipe_t *p = piece->pipe;  
+
+  if((p->want_luminance_mask & DT_DEV_LUMINANCE_MASK_REQUIRED) == 0) return FALSE;
+  if((p->want_luminance_mask & ~DT_DEV_LUMINANCE_MASK_REQUIRED) != mode) return FALSE;
+
+  dt_dev_clear_luminance_mask(p);
+
+  const int width = roi_in->width;
+  const int height = roi_in->height;
+
+  cl_mem out = NULL;
+  float *mask = NULL;
+  const int devid = p->devid;
+
+  mask = dt_alloc_align_float((size_t)width * height);
+  if(mask == NULL) goto error;
+  out = dt_opencl_alloc_device(devid, width, height, sizeof(float));   
+  if(out == NULL) goto error;
+
+  const int program = 31;
+  int kernel_calc_luminance_mask = dt_opencl_create_kernel(program, "out_luminance_mask");
+
+  {
+    size_t sizes[3] = { ROUNDUPWD(width), ROUNDUPHT(height), 1 };
+    dt_opencl_set_kernel_arg(devid, kernel_calc_luminance_mask, 0, sizeof(cl_mem), &out);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_luminance_mask, 1, sizeof(cl_mem), &in);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_luminance_mask, 2, sizeof(int), &width);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_luminance_mask, 3, sizeof(int), &height);
+    const int err = dt_opencl_enqueue_kernel_2d(devid, kernel_calc_luminance_mask, sizes);
+    if(err != CL_SUCCESS) goto error;
+  }  
+
+  {
+    const int err = dt_opencl_read_host_from_device(devid, mask, out, width, height, sizeof(float));
+    if(err != CL_SUCCESS) goto error;
+  }
+
+  p->luminance_mask_data = mask;
+  memcpy(&p->luminance_mask_roi, roi_in, sizeof(dt_iop_roi_t));
+
+  dt_opencl_release_mem_object(out);
+  dt_opencl_free_kernel(kernel_calc_luminance_mask);
```

Comment:
and this should probably be moved into `dt_develop_blend_free_cl_global()`.

Reply:
right again.